### DISCUSSION TITLE
Trying to get Cypress running (without tyler36/ddev-cypress): Missing X server or $DISPLAY

### DISCUSSION THREAD

**Discussion by TravisCarden** [ddev/ddev#6472] [Q&A] [answered]
**tl;dr:** I'm creating a DDEV add-on at https://github.com/TravisCarden/ddev-drupal-xb-dev for configuring a Drupal site for contribution to the Experience Builder initiative. Its major feature is adding interactive Cypress support. I have it working on my machine, but everyone else who has tested it has gotten an X Server error.

> ...without tyler36/ddev-cypress

Okay, please don't hate me. 😛 I know this is supposed to be a solved problem, but https://github.com/tyler36/ddev-cypress doesn't seem to fit my use case, because I need my setup to match the Experience Builder module's Drupal.org GitLab CI implementation. That means that I need to run the Cypress binary it specifies from a directory in the module--not something included in an image. I'm sure I could make it work, but it seems like I would have to override more than I would use.

So here's my problem: I have a complete solution that works perfectly... on my machine. 🤦‍♂️ I can install it via the same instructions as others do, at https://github.com/TravisCarden/ddev-drupal-xb-dev, and it works for me. But they get the following error:

<details>
<summary>Error</summary>

```
$ ddev xb-cypress-open

Cypress failed to start.

This may be due to a missing library or dependency. https://on.cypress.io/required-dependencies

Please refer to the error below for more details.

----------

[2536:0808/191430.352014:ERROR:bus.cc(407)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[2536:0808/191431.055343:ERROR:ozone_platform_x11.cc(240)] Missing X server or $DISPLAY
[2536:0808/191431.055373:ERROR:env.cc(255)] The platform failed to initialize.  Exiting.

----------

Platform: linux-arm64 (Debian - 12)
Cypress Version: 13.12.0
Failed to execute command node_modules/.bin/cypress open --browser electron --project .: exit status 1
Failed to run xb-cypress-open ; error=exit status 1
```

</details>

I assume the difference is on our host machines--since the whole point of Docker is to remove the problem of inconsistencies in the container. 🤷‍♂️ I've tried to think of anything I've done to my host that could be significant--different XQuartz configuration, Cypress cache directories outside the container, environment variables, global node packages--and I can't find it. Does anyone have any ideas?

Related issue: https://github.com/tyler36/ddev-cypress/pull/38

**Comment by rfay** [comment]
Is your machine Intel and the one that doesn't work Apple silicon?

**Comment by tyler36** [comment]
@TravisCarden 
Can you get it working by manually doing the steps? Ie. not using an addon.
Figure out that first, then you'll know what you need to do in the addon.

**Comment by TravisCarden** [comment]
Thank you, both.

@rfay, I believe all the machine's we've tested on have been Apple Silicon--both mine, which works, and everyone else's, which fail.

@tyler36, I've written up a script that performs the steps directly, without using the addon-itself, which I've asked some people to test. The script, with instructions, is here: https://github.com/TravisCarden/ddev-drupal-xb-dev/blob/36b0d6f/tests/debug.sh. It, too, works for me. We'll see if it works for other people, too.

**Comment by chrisfromredfin** [comment]
Jumping in just to try and help by being a total outsider to this process. First thing I noticed:

1) line 82 - the cp of the yml file fails for me. It seems like it's in the wrong path, and can't find it... like it needs to be copying from root, but into var/debug/. I copied it manually then continued manually running the steps...

2) one thing I noticed when accidentally copy/pasting the ddev exec --dir "/var/ww/html/$MODULE_PATH/ui" into my CLI is that for me, on my Mac host, $MODULE_PATH was already defined: `chdir to cwd ("/var/www/html//usr/lib/zsh/5.9/ui") set in config.json failed: no such file or directory: unknown` -- that /usr/lib/zsh/5.9 must be $MODULE_PATH on my host. Something fishy, but whehn I manually did it passing the path, that command worked OK.

3) After getting all that working, I can confirm same as your friends... 🤣
```
Cypress failed to start.

This may be due to a missing library or dependency. https://on.cypress.io/required-dependencies

Please refer to the error below for more details.

----------

[2908:0809/202808.720730:ERROR:bus.cc(407)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[2908:0809/202809.738100:ERROR:ozone_platform_x11.cc(240)] Missing X server or $DISPLAY
[2908:0809/202809.738129:ERROR:env.cc(255)] The platform failed to initialize.  Exiting.

----------

Platform: linux-arm64 (Debian - 12)
Cypress Version: 13.12.0
Failed to execute command node_modules/.bin/cypress open --browser electron --project .: exit status 1
```

Progress thus far: got rid of the dbus error by starting dbus in the web container (ddev ssh, and then 'sudo service start dbus'). 

Still complains that there's no X display (so I tried installing xorg inside the web container, but that didn't help). It seems like what it really needs is "DISPLAY=0 <cmd>" or some way to set that environment var? According to this SO - maybe it needs to be set to the host IP? https://stackoverflow.com/questions/73725613/electron-missing-x-server-or-display

I'm guessing there's some other piece you have that's related to X11 forwarding - maybe your own XQuartz config, etc - something you did that you forgot about??

**Reply by chrisfromredfin** [reply]
I can see now that's what's in the config.drupal-xb-dev.yaml. Anything you need to do with xvfb or xauth - like one-time create an .Xauthority or something?

**Reply by chrisfromredfin** [reply]
I tried install xeyes (through the x11-apps package) inside the web container, just to see if I could get ANY X11 forwarding happening, and that failed also...

```
ddev ssh
cwells@drupal-xb-dev-debug-web:/var/www/html$ xeyes
Error: Can't open display: host.docker.internal:0
```

So $DISPLAY seems set correctly, it just seems to have issues with XQuartz maybe. I noticed you're using the Xquartz maybe bundled with Mac... I'm on Sonoma 14.5 (23F79). You? Them? However, I tried with quartz-wm from Brew, and same issue. Seems to bounce right back from inside the container. Firewall rule preventing outgoing X connections?

**Comment by TravisCarden** [comment]
Thanks, @chrisfromredfin! A total outsider's help is super appreciated. 😛 I saw the `cp` error, too, and fixed that in the repo. `xauth` sounds kind of familiar... I'll look into that.

**Comment by TravisCarden** [comment]
Thanks for the help everyone! We've got everything working now. 🙂

**Reply by rfay** [reply]
Please report the resolution, and link to the working solution, thanks.

**Reply by TravisCarden** [reply, marked as answer]
For several people, it was a case of accidentally skipping some of the documented instructions--specifically, allowing connections from network clients in XQuartz security settings. For one person, we needed to `sleep` for a couple of seconds after opening XQuartz to let it fully start before running Cypress. One person had to change their `$DISPLAY` from `host.docker.internal:0` to `host.orb.internal:0`. (You indicated at https://github.com/tyler36/ddev-cypress/pull/38#discussion_r1714308937 that they should work the same, but one of our testers reported that he had to change it. 🤷‍♂️)

Our complete, working solution can be found at https://github.com/TravisCarden/ddev-drupal-xb-dev, documented as well as I could in the README. They key code components are [`config.drupal-xb-dev.yaml`](https://github.com/TravisCarden/ddev-drupal-xb-dev/blob/develop/config.drupal-xb-dev.yaml) and the host commands, [`xb-cypress-open`](https://github.com/TravisCarden/ddev-drupal-xb-dev/blob/develop/commands/host/xb-cypress-open) and [`xb-cypress-run`](https://github.com/TravisCarden/ddev-drupal-xb-dev/blob/develop/commands/host/xb-cypress-run).

**Reply by rfay** [reply]
Thanks @TravisCarden - Could you say why the @tyler36/ddev-cypress approach didn't work for you?

**Reply by TravisCarden** [reply]
Sure, @rfay. I opted against using `tyler36/ddev-cypress` because it seemed like there was too little overlap in our assumptions and there would've been a lot of overriding and a lot of waste. Specifically, I need to install and run Cypress from the Experience Builder module's `package.json`. So I wouldn't use the container level Cypress install, the `e2e` mount, or the included `cypress-*` commands. That left only the container-level dependencies and configuration, as far as I could tell, which I could just as easily get with a little `web_environment` and `webimage_extra_packages` and eliminate the need for another service/image. I'll be honest: this was my first foray into extending DDEV, and I have no previous experience with Cypress; so I flailed a bit on this, and I could be mistaken. If you have feedback, I'm completely open to it--especially if you have an interest in the matter. From everything I've seen, you're an extremely responsive maintainer, and I really respect that!

**Comment by tyler36** [comment]
Glad to hear you got everything working.

If there's anything in tyler36/ddev-cypress that we can improve on, PRs are alway welcome. Updates to the docs, like troubleshooting or decoding an error message, are espcially helpful for new developers.
